### PR TITLE
Parse Cargo dependencies referencing git repos

### PR DESCRIPTION
**Pull request by rbuch703**

### CODE REVIEW

**Comment by sschuberth** [comment]
Please use [Conventional Commits](https://www.conventionalcommits.org/). The `type` in the title could be either `fix` or `feat` depending on your view; I'd probably say it's a `fix`.

File: plugins/package-managers/cargo/src/main/kotlin/Cargo.kt [hunk 1]
```diff
@@ -297,20 +297,33 @@ private fun parsePackageId(node: JsonNode) =
...
 
+private val GIT_DEPENDENCY_REGEX = Regex("git\\+(https://.*)\\?(?:rev|tag|branch)=.+#([0-9a-zA-Z]+)")
+
 private fun parseSourceArtifact(
     node: JsonNode,
     hashes: Map<String, String>
 ): RemoteArtifact? {
-    if (node["source"].textValueOrEmpty() != "registry+https://github.com/rust-lang/crates.io-index") {
-        return null
+
```
Comment (sschuberth) [inline plugins/package-managers/cargo/src/main/kotlin/Cargo.kt:306]: Nit: Please remove this leading empty line.

**Comment by sschuberth** [comment]
> Please use [Conventional Commits](https://www.conventionalcommits.org/). The `type` in the title could be either `fix` or `feat` depending on your view; I'd probably say it's a `fix`.

Also, the commit message body should contain `Fixes #7372.` in order to auto-close the issue upon merge.

File: plugins/package-managers/cargo/src/main/kotlin/Cargo.kt (hunk 1 above)
Comment (rbuch703) [reply plugins/package-managers/cargo/src/main/kotlin/Cargo.kt]: Done

File: plugins/package-managers/cargo/src/main/kotlin/Cargo.kt [hunk 2]
```diff
@@ -297,20 +297,33 @@ private fun parsePackageId(node: JsonNode) =
...
+    if (source.startsWith("git+https://")) {
+        // Match sources dependencies from git repositories
+        // See https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-git-repositories
+        // for the specification
+        val match = GIT_DEPENDENCY_REGEX.matchEntire(source) ?: return null
+        val (url, hash) = match.destructured
+        return RemoteArtifact(url, Hash.create(hash))
+    }
+
+    return null
```
[the remark this answers is not in the document]
Comment (rbuch703) [reply plugins/package-managers/cargo/src/main/kotlin/Cargo.kt:326]: Done

File: plugins/package-managers/cargo/src/main/kotlin/Cargo.kt [hunk 3]
```diff
@@ -297,20 +297,33 @@ private fun parsePackageId(node: JsonNode) =
...
     }
 
-    val name = node["name"]?.textValue() ?: return null
-    val version = node["version"]?.textValue() ?: return null
-    val url = "https://crates.io/api/v1/crates/$name/$version/download"
-    val id = parseCargoId(node)
-    val hash = Hash.create(hashes[id].orEmpty())
-    return RemoteArtifact(url, hash)
+    if (source.startsWith("git+https://")) {
+        // Match sources dependencies from git repositories
```
[the remark this answers is not in the document]
Comment (rbuch703) [reply plugins/package-managers/cargo/src/main/kotlin/Cargo.kt]: Done

**Review by sschuberth — APPROVED** [review]

**Pull request merged**